TLS (Client did not present a certificate)

Using Haraka for outgoing mails.
Installed Let's Encrypt certificates in tls.ini config file.
both in main and [outbound] sections of config file key=tls_key.pem _(privkey.pem)_
cert=tls_cert.pem _(fullchain.pem)_
Also 3 certificates combined in single file in **config/tls** folder.

Tested the installation with  openssl s_client -connect localhost:25 -starttls smtp
Looks like OK.. Emails are relayed fine.

> [outbound] Trying TLS for domain: receiving_dom.com, host: 222.222.222.222
> [outbound] C: STARTTLS
> [outbound] S: 220 Go ahead\r\n
> [core] client TLS upgrade in progress, awaiting secured.
> [core] client TLS secured.
> [outbound] secured verified=true cipher=ECDHE-RSA-AES128-GCM-SHA256 version=TLSv1/SSLv3 cn=mx.yandex.net organization="Yandex LLC" issuer="Yandex LLC" expires="Nov 27 07:21:48 2019 GMT" fingerprint=B1:1C:5A:9A:93:1D:58:D7:4E:69:75:A0:DC:9E:9E:82:F0:46:75:8F
> .................
> [outbound]  delivered file=156 domain=receiving_dom.com host=mx.yandex.net ip=222.222.222.222 port=25 mode=SMTP tls=Y auth=N response="Ok: queued on mail.yandex.net as 7346831" delay=3.875 fails=0 rcpts=1/0/0

But the receiving mail server adds following header in the mail.
```
Received: from sending_dom.com (sending_dom.com [111.111.111.111])
	by mail.yandex.net (nwsmtp/Yandex) with ESMTPS id se0m3ws
	Sun, 01 Sep 2019 17:00 +0300
	(using TLSv1.2 with cipher ECDHE-RSA-AES128-GCM-SHA256 (128/128 bits))
	(Client certificate not present)
```
Similar certificate warning received from other servers too.
```
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 256/256 bits)
	(Client did not present a certificate)
	by CMGW with ESMTP
```
This problem did not happed with self signed certificates.
### system info
Haraka | Haraka.js — Version: 2.8.24
 --- | :--- 
Node | v10.16.2
OS | Linux 4.19
openssl | OpenSSL 1.0.2r  
### Expected behavior

### Observed behavior

### Steps to reproduce

I ran into a curious issue today. Haraka was unable to deliver to mail to a specific mail server. Investigating further, it's because the TLS connection simply does not finish. However, swaks just works fine.

```
[DEBUG] [-] [core] client TLS upgrade in progress, awaiting secured.
[ERROR] [8B8E0D82-25CB-4592-A1B0-EDE04C61C9AA.1.1] [outbound] Ongoing connection failed to xxxxxx:25 : Error: Client network socket disconnected before secure TLS connection was established
```

On further investigation, I found that this the connection does not finish when `key` and `cert` are sent. If I remove the key and cert for outbound tls, the connection works fine. Note that key and cert are valid LE certs. Also note that this issue only happens with a specific mail server (!).

@msimerson @celesteking I don't know how mTLS works but how is it possible that connection only fails when presenting valid certs? In general, is it required to send certs at all for outbound ? Can https://github.com/haraka/Haraka/blob/3e779b00b554e54fd8b1151e7af0734961007f2b/tls_socket.js#L709 be removed below? I thought sending certs is only required if the server requests certs. Why would server request client certs in SMTP world? I am referring to the code below:

```
        if (exports.tls_valid) {
            options = Object.assign(options, certsByHost['*']);
        }
```

Finally, I guess http://haraka.github.io/plugins/tls/#inbound-specific-configuration is obsolete ? AFAICT, there is no such thing as outbound and inbound section anymore in the code. I can submit a PR for this.

--- a/Changes.md
+++ b/Changes.md
@@ -3,9 +3,11 @@
 
 ### Changes
 
+- tls: change default to NOT send TLS client certs #2902
+- dep: redis is now a dependency #2896
 - use address-rfc2821 2.0.0
-- drop support for node 10
-- http: use CDN for bootstrap/jquery, drop bower
+- http: use CDN for bootstrap/jquery, drop bower #2891
+- drop support for node 10  #2890
 
 ### New features
 
@@ -13,7 +15,9 @@
 
 ### Fixes
 
-- 
+- bounce: correctly set fail recipients #2901
+- bounce: correctly set bounce recipients #2899
+
 
 ## 2.8.27 - 2021-01-05
 

--- a/config/tls.ini
+++ b/config/tls.ini
@@ -26,6 +26,10 @@
 ; requireAuthorized[]=465
 ; requireAuthorized[]=587
 
+; send client certificate(s). If you use this setting and value it, report
+; your use case at https://github.com/haraka/Haraka/issues/2693
+; mutual_tls=false
+
 
 [redis]
 ; options in this block require redis to be enabled in config/plugins.
@@ -48,11 +52,24 @@
 ; TLS NO-GO Inbound expiry time in seconds
 ; disable_inbound_expiry = 3600
 
+
 ; no_tls_hosts - disable TLS for servers with broken TLS. (applies to inbound only)
 [no_tls_hosts]
 ; 127.0.0.1
 ; 192.168.1.1
 ; 172.16.0.0/16
+
+
+; hosts that require us to present a cert signed by a CA we both trust
+[mutual_auth_hosts]
+;travel.state.gov                     ; use default TLS cert
+;xo.huggable.gov=special.my-tld.com   ; specify cert by CN
+
+
+; these hosts request mutual TLS and reject our TLS certificate
+[mutual_auth_hosts_exclude]
+;bofh.no-such-agency.gov
+
 
 [outbound]
 ; key=tls_key.pem

--- a/tests/tls_socket.js
+++ b/tests/tls_socket.js
@@ -132,9 +132,12 @@ exports.load_tls_ini2 = {
                     requestOCSP: false,
                     // enableOCSPStapling: false,
                     requireAuthorized: [],
+                    mutual_tls: false,
                 },
                 redis: { disable_for_failed_hosts: false },
-                no_tls_hosts: {}
+                no_tls_hosts: {},
+                mutual_auth_hosts: {},
+                mutual_auth_hosts_exclude: {},
             });
         test.done();
     },
@@ -153,9 +156,12 @@ exports.load_tls_ini2 = {
                 ciphers: 'ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-SHA384',
                 minVersion: 'TLSv1',
                 requireAuthorized: [2465, 2587],
+                mutual_tls: false,
             },
             redis: { disable_for_failed_hosts: false },
             no_tls_hosts: {},
+            mutual_auth_hosts: {},
+            mutual_auth_hosts_exclude: {},
             outbound: {
                 key: 'outbound_tls_key.pem',
                 cert: 'outbound_tls_cert.pem',

--- a/tls_socket.js
+++ b/tls_socket.js
@@ -234,12 +234,15 @@ exports.load_tls_ini = (opts) => {
             '-main.rejectUnauthorized',
             '+main.honorCipherOrder',
             '-main.requestOCSP',
+            '-main.mutual_tls',
         ]
     }, () => {
         tlss.load_tls_ini();
     });
 
-    if (!cfg.no_tls_hosts) cfg.no_tls_hosts = {};
+    if (cfg.no_tls_hosts === undefined) cfg.no_tls_hosts = {};
+    if (cfg.mutual_auth_hosts === undefined) cfg.mutual_auth_hosts = {};
+    if (cfg.mutual_auth_hosts_exclude === undefined) cfg.mutual_auth_hosts_exclude = {};
 
     if (cfg.main.enableOCSPStapling !== undefined) {
         log.logerror('deprecated setting enableOCSPStapling in tls.ini');
@@ -688,6 +691,11 @@ function createServer (cb) {
     return server;
 }
 
+function getCertFor (host) {
+    if (host && certsByHost[host]) return certsByHost[host];
+    return certsByHost['*'];  // the default TLS cert
+}
+
 function connect (port, host, cb) {
     let conn_options = {};
     if (typeof port === 'object') {
@@ -708,7 +716,22 @@ function connect (port, host, cb) {
         cryptoSocket.removeAllListeners('data');
 
         if (exports.tls_valid) {
-            options = Object.assign(options, certsByHost['*']);
+            /* SUNSET notice: code added 2021-01. We've changed the default to not
+               send TLS client certificates. The mutual_tls flag switches them back
+               on. If no need for these settings surfaces in 2 years, nuke this block
+               of code. If you care about these options, create a PR removing this
+               comment. See #2693.
+            */
+            if (exports.cfg === undefined) exports.load_tls_ini();
+            if (exports.cfg.mutual_auth_hosts[host]) {
+                options = Object.assign(options, getCertFor(exports.cfg.mutual_auth_hosts[host]));
+            }
+            else if (exports.cfg.mutual_auth_hosts_exclude[host]) {
+                // send no client cert
+            }
+            else if (exports.cfg.main.mutual_tls) {
+                options = Object.assign(options, getCertFor(host));
+            }
         }
         options.socket = cryptoSocket;
 

